File: app/src/test/java/com/moisesborges/tvaddict/episodes/EpisodesPresenterTest.java
package com.moisesborges.tvaddict.episodes;

import com.moisesborges.tvaddict.data.ShowsRepository;
import com.moisesborges.tvaddict.models.Episode;
import com.moisesborges.tvaddict.models.Season;
import com.moisesborges.tvaddict.models.Show;
import com.moisesborges.tvaddict.utils.MockShow;
import com.moisesborges.tvaddict.utils.RxJavaTestConfig;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;
import org.mockito.internal.creation.bytebuddy.MockAccess;

import java.util.List;
import java.util.concurrent.Callable;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.functions.Function;

import static org.hamcrest.core.Is.is;
import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

/**
 * Created by Moisés on 21/04/2017.
 */
public class EpisodesPresenterTest {
    
    @Mock
    EpisodesView mEpisodesView;
    @Mock
    ShowsRepository mShowsRepository;
    
    private EpisodesPresenter mEpisodesPresenter;
    private Show mShow;
    
    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);

        mShow = MockShow.newMockShowInstance();
        mShow.setEmbedded(MockShow.newEmbeddedInstance());
        
        mEpisodesPresenter = new EpisodesPresenter(new RxJavaTestConfig(), mShowsRepository);
    }

    @Test
    public void shouldDisplayEpisodesOfRightSeasonAndLoadSavedShow() throws Exception {
        mEpisodesPresenter.bindView(mEpisodesView);

        when(mShowsRepository.getSavedShow(mShow.getId()))
                .thenReturn(Single.just(mShow));

        Season season = mShow.getSeasons().get(0);
        List<Episode> episodes = mShow.getEpisodes()
                .stream()
                .filter(episode -> episode.getSeason().equals(season.getNumber()))
                .collect(Collectors.toList());

        mEpisodesPresenter.loadEpisodes(season.getId(), mShow);
        verify(mEpisodesView).setShow(mShow);
        verify(mEpisodesView).displayEpisodes(episodes);
    }

    @Test
    public void shouldDisplayEpisodesOfRightSeasonWithouLoadShow() throws Exception {
        mEpisodesPresenter.bindView(mEpisodesView);

        when(mShowsRepository.getSavedShow(mShow.getId()))
                .thenReturn(Single.just(Show.NOT_FOUND));

        Season season = mShow.getSeasons().get(0);
        List<Episode> episodes = mShow.getEpisodes()
                .stream()
                .filter(episode -> episode.getSeason().equals(season.getNumber()))
                .collect(Collectors.toList());

        mEpisodesPresenter.loadEpisodes(season.getId(), mShow);
        verify(mEpisodesView, never()).setShow(mShow);
        verify(mEpisodesView).displayEpisodes(episodes);
    }

    @Test
    public void shouldSetEpisodeStatusAsSeen() throws Exception {
        mEpisodesPresenter.bindView(mEpisodesView);

        Mockito.when(mShowsRepository.updateShow(mShow)).thenReturn(Single.just(mShow));

        Episode episode = mShow.getEpisodes().get(0);
        episode.setWatched(false);

        mEpisodesPresenter.changeEpisodeSeenStatus(episode, mShow);

        assertThat(episode.wasWatched(), is(true));
        verify(mShowsRepository).updateShow(mShow);
        verify(mEpisodesView).refreshEpisode(episode);
    }

    @Test
    public void shouldSetEpisodeStatusAsUnseen() throws Exception {
        mEpisodesPresenter.bindView(mEpisodesView);

        Mockito.when(mShowsRepository.updateShow(mShow)).thenReturn(Single.just(mShow));

        Episode episode = mShow.getEpisodes().get(0);
        episode.setWatched(true);

        mEpisodesPresenter.changeEpisodeSeenStatus(episode, mShow);

        assertThat(episode.wasWatched(), is(false));
        verify(mShowsRepository).updateShow(mShow);
        verify(mEpisodesView).refreshEpisode(episode);
    }
}

File: app/src/test/java/com/moisesborges/tvaddict/episodes/EpisodesPresenterTest.kt
package com.moisesborges.tvaddict.episodes

import com.moisesborges.tvaddict.data.ShowsRepository
import com.moisesborges.tvaddict.models.Episode
import com.moisesborges.tvaddict.models.Show
import com.moisesborges.tvaddict.utils.DateUtils
import com.moisesborges.tvaddict.utils.MockShow
import com.moisesborges.tvaddict.utils.RxJavaTestConfig
import io.reactivex.Single
import org.hamcrest.core.Is.`is`
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.MockitoAnnotations.initMocks
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Created by Moisés on 21/04/2017.
 */
class EpisodesPresenterTest {

    @Mock
    internal lateinit var episodesView: EpisodesView
    @Mock
    internal lateinit var showsRepository: ShowsRepository

    private lateinit var episodesPresenter: EpisodesPresenter
    private lateinit var show: Show

    @Before
    fun setUp() {
        initMocks(this)

        show = MockShow.newFullMockShowInstance()
        episodesPresenter = EpisodesPresenter(RxJavaTestConfig(), showsRepository)
    }

    @Test
    fun shouldDisplayEpisodesOfRightSeasonAndLoadSavedShow() {
        episodesPresenter.bindView(episodesView)

        `when`(showsRepository.getSavedShow(show.id)).thenReturn(Single.just(show))

        val season = show.seasons.first()
        val episodes = show.episodes
                .filter { episode -> episode.season == season.number }

        episodesPresenter.loadEpisodes(season.id, show)
        verify<EpisodesView>(episodesView).setShow(show)
        verify<EpisodesView>(episodesView).displayEpisodes(episodes)
    }

    @Test
    fun shouldDisplayEpisodesOfRightSeasonWithouLoadShow() {
        episodesPresenter.bindView(episodesView)

        `when`(showsRepository.getSavedShow(show.id!!))
                .thenReturn(Single.just(Show.NOT_FOUND))

        val season = show.seasons.first()
        val episodes = show.episodes
                .filter { episode -> episode.season == season.number }

        episodesPresenter.loadEpisodes(season.id, show)
        verify<EpisodesView>(episodesView, never()).setShow(show)
        verify<EpisodesView>(episodesView).displayEpisodes(episodes)
    }

    @Test
    fun shouldSetEpisodeStatusAsSeen() {
        episodesPresenter.bindView(episodesView)

        `when`(showsRepository.updateShow(show)).thenReturn(Single.just(show))

        val episode = show.episodes.first()
        episode.setWatched(false)

        episodesPresenter.changeEpisodeSeenStatus(episode, show)

        assertThat(episode.wasWatched(), `is`(true))
        verify<ShowsRepository>(showsRepository).updateShow(show)
        verify<EpisodesView>(episodesView).refreshEpisode(episode)
    }

    @Test
    fun shouldNotSetEpisodeStatusAsSeen() {
        episodesPresenter.bindView(episodesView)

        `when`(showsRepository.updateShow(show)).thenReturn(Single.just(show))

        val episode = show.episodes.first()

        episode.setWatched(false)

        val todayPlusOneHour = Date(Date().time + TimeUnit.HOURS.toMillis(1))
        changeEpisodeAirdateAndAirtime(episode, todayPlusOneHour)

        episodesPresenter.changeEpisodeSeenStatus(episode, show)

        assertThat(episode.wasWatched(), `is`(false))
        verify(showsRepository, never()).updateShow(show)
        verify(episodesView, never()).refreshEpisode(episode)
    }

    private fun changeEpisodeAirdateAndAirtime(episode: Episode, date: Date) {
        episode.airdate = DateUtils.dateToAirdate(date)
        episode.airtime = DateUtils.dateToAirtime(date)
    }

    @Test
    fun shouldSetEpisodeStatusAsUnseen() {
        episodesPresenter.bindView(episodesView)

        `when`(showsRepository.updateShow(show)).thenReturn(Single.just(show))

        val episode = show.episodes.first()
        episode.setWatched(true)

        episodesPresenter.changeEpisodeSeenStatus(episode, show)

        assertThat(episode.wasWatched(), `is`(false))
        verify<ShowsRepository>(showsRepository).updateShow(show)
        verify<EpisodesView>(episodesView).refreshEpisode(episode)
    }

    @Test
    fun shouldSetAllSeasonEpisodesStatusAsSeen() {
        episodesPresenter.bindView(episodesView)

        `when`(showsRepository.updateShow(show)).thenReturn(Single.just(show))

        val season = show.seasons.first()
        val episodes = show.episodes
                .filter { episode -> episode.season == season.number }

        episodesPresenter.setAllEpisodesStatusAsSeen(season.number, show)

        episodes.forEach { episode -> assertThat(episode.wasWatched(), `is`(true)) }
        verify(showsRepository).updateShow(show)
        verify(episodesView).refreshAllEpisodes()
    }

    @Test
    fun shouldNotSetAllSeasonEpisodesStatusAsSeen() {
        episodesPresenter.bindView(episodesView)

        `when`(showsRepository.updateShow(show)).thenReturn(Single.just(show))

        val season = show.seasons.first()
        val episodes = show.episodes
                .filter { episode -> episode.season == season.number }

        val lastTwoEpisodes = episodes.takeLast(2)

        val todayPlusOneHour = Date(Date().time + TimeUnit.HOURS.toMillis(1))
        lastTwoEpisodes.forEach { episode -> changeEpisodeAirdateAndAirtime(episode, todayPlusOneHour) }

        episodesPresenter.setAllEpisodesStatusAsSeen(season.number, show)

        episodes.forEach { episode -> val trueIfAired = !lastTwoEpisodes.contains(episode)
            assertThat(episode.wasWatched(), `is`(trueIfAired)) }
        verify(showsRepository).updateShow(show)
        verify(episodesView).refreshAllEpisodes()
    }

    @Test
    fun shouldSetAllSeasonEpisodesStatusAsUnseen() {
        episodesPresenter.bindView(episodesView)

        `when`(showsRepository.updateShow(show)).thenReturn(Single.just(show))

        val season = show.seasons.first()
        val episodes = show.episodes
                .filter { episode -> episode.season == season.number }

        episodesPresenter.setAllEpisodesStatusAsUnseen(season.number, show)

        episodes.forEach { episode -> assertThat(episode.wasWatched(), `is`(false)) }
        verify(showsRepository).updateShow(show)
        verify(episodesView).refreshAllEpisodes()
    }
}